src/services/transaction_watcher_service.ts

```diff
@@ -0,0 +1,117 @@
+            await this._findTransactionStatusAndUpdateAsync(tx);
+        }
+    }
+    private async _findTransactionStatusAndUpdateAsync(txEntity: TransactionEntity): Promise<TransactionEntity> {
+        // TODO(oskar) - LessThanOrEqual and LessThan do not work on dates in
+        // TypeORM queries, ref: https://github.com/typeorm/typeorm/issues/3959
+        const now = new Date();
+        const isExpired = txEntity.expectedAt <= now;
+        try {
+            const txInBlockchain = await this._web3Wrapper.getTransactionByHashAsync(txEntity.hash);
+            if (txInBlockchain !== undefined && txInBlockchain !== null && txInBlockchain.hash !== undefined) {
+                if (txInBlockchain.blockNumber !== null) {
+                    logger.trace({
+                        message: `a transaction with a ${txEntity.status} status is already on the blockchain, updating status to TransactionStates.Confirmed`,
+                        hash: txInBlockchain.hash,
+                    });
+                    txEntity.status = TransactionStates.Confirmed;
+                    txEntity.blockNumber = txInBlockchain.blockNumber;
+                    await this._transactionRepository.save(txEntity);
+                    await this._abortTransactionsWithTheSameNonceAsync(txEntity);
+                    return txEntity;
+                    // Checks if the txn is in the mempool but still has it's status set to Unsubmitted or Submitted
+                } else if (!isExpired && txEntity.status !== TransactionStates.Mempool) {
+                    logger.trace({
+                        message: `a transaction with a ${txEntity.status} status is pending, updating status to TransactionStates.Mempool`,
+                        hash: txInBlockchain.hash,
+                    });
+                    txEntity.status = TransactionStates.Mempool;
+                    return this._transactionRepository.save(txEntity);
+                } else if (isExpired) {
+                    // NOTE(oskar): we currently cancel all transactions that are in the
+                    // "stuck" state. A better solution might be to unstick
+                    // transactions one by one and observing if they unstick the
+                    // subsequent transactions.
+                    txEntity.status = TransactionStates.Stuck;
+                    return this._transactionRepository.save(txEntity);
+                }
+            }
+        } catch (err) {
+            if (err instanceof TypeError) {
+                // HACK(oskar): web3Wrapper.getTransactionByHashAsync throws a
+                // TypeError if the Ethereum node cannot find the transaction
+                // and returns NULL instead of the transaction object. We
+                // therefore use this to detect this case until @0x/web3-wrapper
+                // is fixed.
+                if (isExpired) {
+                    txEntity.status = TransactionStates.Dropped;
```

Comment:
How certain are we that case is actually `Dropped` and that we're not just talking to a different node behind a load balancer with a different view of the mempool?

> returns NULL instead of the transaction object

I was under the impression this was always geth's behaviour, and parity would return the tx object for unconfirmed transactions. But it's been so long

Reply:
The transaction will only be considered "dropped" when it is expired AND is not present in the mempool according to the RPC provider.

I guess we can never be sure about nodes with a different views of the mempool, only minimize that risk by:
- grabbing the transaction using the same provider we submit it to (which I agree, does not guarantee that the transaction will be reflected in the mempool).
- using multiple RPC providers (alchemy, infura, our self-hosted node) and consider that dropped when all of them agree on that.

Is there a better way?

>  I was under the impression this was always geth's behaviour, and parity would return the tx object for unconfirmed transactions. But it's been so long

That is the case, but the [marshaller](https://github.com/0xProject/0x-monorepo/blob/development/packages/web3-wrapper/src/web3_wrapper.ts#L253) in the web3 wrapper does not actually check if the return is null, and throws a `TypeError`.